### ISSUE TITLE
unnecessary_wraps is a bad default lint

### ISSUE THREAD

**Issue by mitsuhiko** [rust-lang/rust-clippy#6726]
Lint name: `unnecessary_wraps`

The newly introduced `unnecessary_wraps` lint is one I now need to disable everywhere it has popped up so far as a warning. The reason for this is that we use `serde(default = func)` a lot and when you have a nullable type you need to return a `Some(value)` from that function.

I would make an argument that this lint should not be in the list of default lints because of there being a lot of legitimate cases where one does need "unnecessary wrapping".

**Comment by foresterre** [comment]
I also think the `unnecessary_wraps` should not be enabled by default.

In my use case, some functions return a `Result<(), Error>`, which strictly speaking is unnecessary, but makes the code of the caller more readable.

Take for example the following code where the `unnecessary_wraps` lint pops up:

```rust
fn process_instruction(&mut self, instruction: Instr) -> Result<(), SicImageEngineError> {
    match instruction {
        Instr::Operation(op) => self.process_operation(op),
        Instr::EnvAdd(item) => self.insert_env(*item),
        Instr::EnvRemove(key) => self.remove_env(*key),
    }
}
```

Here, `self.insert_env` and `self.remove_env`, do not result in a failure case, but return a `Ok(())` regardless, to improve the readability of the `process_instruction` from which they are called.
 
The `self.process_operation(op)` can result in a SicImageEngineError so does not unnecessary wrap.

After fixing the code to resolve this linter issue, it becomes:

```rust
fn process_instruction(&mut self, instruction: &Instr) -> Result<(), SicImageEngineError> {
    match instruction {
        Instr::Operation(op) => self.process_operation(op),
        Instr::EnvAdd(item) => {
            self.insert_env(*item);
            Ok(())
        }
        Instr::EnvRemove(key) => {
            self.remove_env(*key);
            Ok(())
        }
    }
}
```

Here, the complexity of the `Instr::EnvAdd` and `Instr::EnvRemove` match arms are increased and consequently, the complete function is less readable. 


<details>
  <summary>Another example</summary>
  
	  ```rust
	  impl Action for UpdateManifest<'_> {
	      fn run(&mut self, args: &PublishWorkspace) -> anyhow::Result<()> {
	          if args.dry_run {
	        	  // This fn returns Ok(()) "unnecessarily", because it will never return an error,
	              // but fixing the lint decreases code readability 
	              dry_update_dependency_version(self.package, args.version())
	          } else {
	        	  // This fn can actually fail
	              live_update_dependency_version(self.package, args.version())
	          }
	      }
	  }
	  ```
</details>

**Comment by mitsuhiko** [comment]
It's also add that you would have to refactor your API just because the code in the current implementation might not return an error.

**Comment by glts** [comment]
Another example where this lint may not be desirable: I have a small parser [here](https://gitlab.com/glts/viaspf/-/blob/0.3.0/src/record/parse.rs#L175) where all parsing functions have a similar signature (very much like nom parsers):

```
fn some_object(input: &str) -> Option<(&str, SomeObject)>
```

Sometimes the parsing function is infallible and always returns `Some`. As a design, and for callers, I believe it is more consistent to keep `Option` as the return type.

**Comment by llogiq** [comment]
Even with downgrading the lint, a check to not lint exported items would be useful.
